### PR TITLE
CTRL-C event kills the application

### PR DESCRIPTION
**Pull request by cbucher**
git-cmd.exe has a bad behavior when it receives CTRL-C event. A command line shell must handle the events.
For example, CTRL-C should be used to terminate the current command and not to terminate the shell himself.

When git-cmd.exe is launched, git-cmd.exe launches a child process (for example cmd.exe).
git-cmd.exe monitors the child process. When the child process exits, git-cmd.exe exits too.

When the user presses CTRL-C, the Windows console receives Windows messages for WM_KEYDOWN/WM_KEYUP.
Then the TranslateMessage function generates a CTRL-C event.

The CTRL-C event is handled by the Control Handler Function.
When a CTRL_BREAK_EVENT, CTRL_LOGOFF_EVENT, or CTRL_SHUTDOWN_EVENT signal is received, the control handler returns FALSE. Doing this causes the signal to be passed to the next control handler function. If no other control handlers have been registered or none of the registered handlers returns TRUE, the default handler will be used, resulting in the process being terminated.

cmd.exe Control Handler function terminates the current command and returns FALSE.
git-cmd.exe has no Control Handler Function registered. So git-cmd.exe is terminated.

When git-cmd.exe is used with programs like ConsoleZ, ConsoleZ injects a dll in git-cmd.exe and monitors this process.
When the user presses CTRL-C, git-cmd is terminated and so the ConsoleZ's tab is closed.

To fix this behavior, we can simply register a Control Handler function that returns TRUE (for all events). In this case the only way to exit this program is the launched command end.

Signed-off-by: Christophe Bucher Developer <christophe.bucher@laposte.net>

### CODE REVIEW

**Comment by bviktor** [comment]
I've tested this patch, it seems to fix the issue and no regressions so far.

**Comment by dscho** [comment]
This is almost the solution I was looking for. I will reformat the commit message to make it comply with Git's standards, no need to worry about that.

The bigger worry is that there is a chance that we need to be able to *disable* the new behavior, most likely by introducing an appropriate command-line option. `git-cmd.exe` has to cater to more than just ConsoleZ.

I will still need to test whether everything is working alright (I could imagine that Ctrl+C is no longer handled correctly when trying to interrupt a runaway executable spitting out millions of lines).

However, I tested that the current `git-cmd.exe`, i.e. the one *without* this patch works correctly when I double-click it in the Explorer and press Ctrl+C.

**Comment by bviktor** [comment]
> However, I tested that the current git-cmd.exe, i.e. the one without this patch works correctly when I double-click it in the Explorer and press Ctrl+C.

@dscho I hope you do realize that `git-cmd.exe` on its on was never an issue. The issue is with `git-cmd.exe --command=usr\bin\bash.exe --login -i`.

**Comment by dscho** [comment]
> I hope you do realize that git-cmd.exe on its own was never an issue. The issue is with git-cmd.exe --command=usr\bin\bash.exe --login -i.

I just created a shortcut, launched it, and it still does exactly what I want it to do: if I press Ctrl+C at the Bash prompt, it cancels the current command line and gives me a new one. If I run `cat` and press Ctrl+C, that program is interrupted. Even if I run `git rev-list HEAD` -- which is a Win32 executable, not an MSys2 one -- and press Ctrl+C, that program is interrupted as well.

So yes, I think I realize how you run `git-cmd`. It's just that doing so outside of ConsoleZ seems to work without any problem.

**Comment by bviktor** [comment]
@dscho press Win+R (Run command), start `git-cmd.exe --command=usr\bin\bash.exe --login -i`, press Ctrl+C, and check out your taskmgr, git-cmd will be gone, even tho your prompt will be retained.

Press Win+R, start cmd.exe, `git-cmd.exe --command=usr\bin\bash.exe --login -i`, press Ctrl+C twice, and on top of git-cmd being gone, you'll only get `More ?` messages upon keypresses.

**Comment by dscho** [comment]
Unfortunately, my concerns were legitimate: if I run the patches `git-cmd.exe` and try to interrupt a long-running `git rev-list HEAD`, it simply freezes. With no way to recover other than calling the Task Manager, guessing the correct `git.exe` and terminating the process (with two frightening warnings).

**Comment by dscho** [comment]
> press Win+R (Run command), start git-cmd.exe --command=usr\bin\bash.exe --login -i

FWIW I get the exact same freezing behavior with the patches `git-cmd.exe` when I call it that way rather than via a newly-created shortcut.

**Comment by dscho** [comment]
@bviktor @cbucher as I do not have ConsoleZ, could you please test whether you also experience the freezing behavior?

**Comment by bviktor** [comment]
@dscho sure, just give me a repo and the exact command you're running :)

**Comment by bviktor** [comment]
Actually, I've found a very simple way of testing freezes: just start `git-cmd.exe --command=usr\bin\bash.exe --login -i` and press Ctrl+C constantly. It will freeze within a few seconds.

Here's the funny thing: if you let it run for like a minute, it'll eventually return the prompt. So I'm not quite sure the freeze you encountered is caused by the patch.

Now the bad news: this happens both with _and_ without the patch.

**Comment by dscho** [comment]
@cbucher I could imagine that something like this is actually what you need:

```diff
diff --git a/compat/win32/git-wrapper.c b/compat/win32/git-wrapper.c
index 396a4ee..27b7777 100644
--- a/compat/win32/git-wrapper.c
+++ b/compat/win32/git-wrapper.c
@@ -535,6 +535,8 @@ int main(void)
 			si.dwFlags |= STARTF_USESHOWWINDOW;
 			si.wShowWindow = SW_SHOW;
 		}
+		if (wait)
+			creation_flags |= CREATE_NEW_PROCESS_GROUP;
 		br = CreateProcess(/* module: null means use command line */
 				exep,
 				cmd,  /* modified command line */
```

**Comment by bviktor** [comment]
@dscho for me all the difference this patch makes is that the freeze kicks in after 10 seconds instead of just 1.

**Comment by bviktor** [comment]
RC3 introduced the redirectors. This means that I can use just `bin\bash.exe --login -i` and git commands will still be available.

Now the funny thing is that even tho git-cmd.exe is no longer involved,

* Ctrl+C still closes the tab in ConsoleZ
* two consecutive Ctrl+C presses still produce `More ?` messages when started from CMD
* it still won't close upon Ctrl+C when started via Win+R, but then spamming Ctrl+C still causes freezes 

So basically, `bin\bash.exe --login -i` produces the exact same symptoms as `git-cmd.exe --command=usr\bin\bash.exe --login -i`. Yeesh.

So after all, it may not be git-cmd's fault at all. At this point, it seems more like a bash problem.

**Comment by dscho** [comment]
> So after all, it may not be git-cmd's fault at all. At this point, it seems more like a bash problem.

No, it is still the Git wrapper's problem (or ConsoleZ's, still unsure), because the redirector is simply an adapted copy of the Git wrapper, just like `git-cmd.exe`, except that it executes a different command-line by default.

**Comment by dscho** [comment]
@bviktor did you have a chance to test [the `CREATE_NEW_PROCESS_GROUP` idea](https://github.com/git-for-windows/git/pull/205#issuecomment-113191548) yet?

**Comment by bviktor** [comment]
> @bviktor did you have a chance to test the CREATE_NEW_PROCESS_GROUP idea yet?

Sure, [I already tried that](https://github.com/git-for-windows/git/pull/205#issuecomment-113200912).

**Comment by dscho** [comment]
Hmm. This makes me believe that there is *another*, completely separate issue going on there, that is not helped by the ctrl handler nor the new process group?

**Comment by bviktor** [comment]
Should I open a separate ticket about Ctrl+C spam freezing?

**Comment by dscho** [comment]
@bviktor I think this should be a separate ticket, because the symptom is already different, therefore I suspect the cause to be different, too.

If this still happens even with the ctrl handler patch of @cbucher, I strongly suspect that it is a ConsoleZ issue, either. In any case, it will require quite a lot of effort to figure out, I am afraid.

**Comment by cbucher** [comment]
>  I strongly suspect that it is a ConsoleZ issue
ConsoleZ issue for an issue you can reproduce without ConsoleZ?!?
ahem...

**Comment by bviktor** [comment]
Lol, I believe it was a typo :D

**Comment by dscho** [comment]
> > I strongly suspect that it is a ConsoleZ issue
> 
> ConsoleZ issue for an issue you can reproduce without ConsoleZ?!?
> ahem...

Well, given that I *cannot* reproduce that freeze after 10 seconds, I assume that it is a ConsoleZ issue.

**Comment by bviktor** [comment]
@dscho this happens with standard Git2, zero ConsoleZ involved. But I'll have to check it on a Win8 comp, too. If you can't repro, this may be Win10-specific then.

**Comment by dscho** [comment]
@cbucher BTW I think that we should introduce a new command-line option to install that ctrl handler and not do that by default, seeing as it would cause problems with MinTTY if we do so. I trust you agree that this is a good compromise.

@bviktor I cannot reproduce. Git2, zero ConsoleZ. No freeze. I also got that exact same behavior on 8.1, 7 and XP.

**Comment by cbucher** [comment]
@dscho Hi
My only concern is to limit the number of issues opened in ConsoleZ's about git-cmd.exe and git-bash.exe.

git-cmd.exe is killed with CTRL-C. Any program monitoring the launched process is impacted.

git-bash.exe launch mintty configured to look like. So I must explained to people that I can attach to a non CLI application.

When bash is launched from git-cmd.exe, spamming CTRL-C (simply hold down the keys) makes the application hang. I don't know why but it seems:
- bash (A) launches another process bash (B)
- this process (B) terminates (by CTRL-C ?) 
- bash (A) is waiting for something from the killed process (B)...
I don't have time for reading bash source code.

And all these bugs can be reproduced without ConsoleZ.

With git.cmd, I didn't have tickets.
With git-cmd.exe, I will have tickets.
So this is a regression...

What is the added value of these new executables?

**Comment by dscho** [comment]
> What is the added value of these new executables?

The added value is that we can finally address issues such as making Git Bash pinnable, with a nice icon, with proper environment variable support (we can interact with the entire Win32 API now).

Since this ticket seems to be the same as #227, and the latter ticket has an MCVE that does not require ConsoleZ, I will close this ticket now.

**Comment by cbucher** [comment]
> The added value is that we can finally address issues such as making Git Bash pinnable, with a nice icon, with proper environment variable support (we can interact with the entire Win32 API now).

This can be done by properly creating a shortcut. No?

**Comment by cristovaov** [comment]
For ControlZ users (and perhaps others using a 'console wrapper') with a Ctrl+C problem:
set your tab shell path to `C:\Program Files\Git\usr\bin\sh.exe --login -i` (my settings screenshot below)

My understanding from looking at my processes in Task Manager is that `git-cmd.exe`, `git-bash.exe` and `./bin/*.exe` are console wrappers themselves -- further more simply running these from the shortcuts/file explorer launches a console with I'm assuming your expected behavior of Ctrl+C 
It is thus in my opinion unnecessary to call them if your using your own wrapper (in my case, ConsoleZ) and I am considering this more of an issue of personal settings instead of a GfW issue.

bash setting consolez

**Comment by dscho** [comment]
> > The added value is that we can finally address issues such as making Git Bash pinnable, with a nice icon, with proper environment variable support (we can interact with the entire Win32 API now).
> 
> This can be done by properly creating a shortcut. No?

No. The target of a shortcut is always an absolute path (not possible for us; we also ship portable Git), and there is no way to call a shortcut from a script and pass parameters. Sorry, not a solution.

**Comment by dscho** [comment]
> set your tab shell path to C:\Program Files\Git\usr\bin\sh.exe --login -i

@cristovaov this does not work correctly, unfortunately, because you need to configure MSys2 with environment variables, most notably by setting `MSYSTEM`.

**Comment by Metallium** [comment]
@dscho using `usr\bin\sh.exe --login -i` instead of `bin\sh.exe --login -i` resolves issue with disposing on CTRL+C, but $MSYSTEM becomes unset.
What do we sacrifice and what won't work correctly in that case?

**Comment by dscho** [comment]
`MSYSTEM` was only one example... You can see what the wrapper does here: https://github.com/git-for-windows/git/blob/master/compat/win32/git-wrapper.c

**Comment by cristovaov** [comment]
It was something I had noticed Friday as I was posting the comment. 

@Metallium you can resolve $MSYSTEM by setting it in the Environment option of Tab settings (ref: screenschot previous comment) -- Variable: `MSYSTEM` Value: `MINGW64` 
The main drawback of $MSYSTEM being unset I have encountered was in Vim where time and date will be incorrect: 
vim-time-date

Thanks for pointing out the wrapper file. While enlightening I wasn't able to fully grasp the added benefits of this config (and I don't speak the C language...). Is there a dedicated channel where I could send in my questions?

I noticed you provided Vagrant file too, will have a run at it later this week I think. I'm curious about the building process.

**Comment by dscho** [comment]
> you can resolve $MSYSTEM by setting it in the Environment option of Tab settings (ref: screenschot previous comment) -- Variable: MSYSTEM Value: MINGW64

You will also need to set `HOME` and `TERM`, at least. And maybe adjust the `PATH`.

> Is there a dedicated channel where I could send in my questions?

Why, yes! Just follow the *Mailing List* link on the top right of [Git for Windows 2.x' home page](http://git-for-windows.github.io/): http://groups.google.com/group/git-for-windows

> I noticed as well that you provided Vagrant file too, will have a run at it later this week I think. I'm curious about the building process.

Well, the Vagrantfile will build a *Linux* Git, of course. That's what Vagrant does (see also https://github.com/git-for-windows/git/wiki/Vagrant). If you are curious about the building process of Git for Windows, there is quite a bit of information on our wiki:

- https://github.com/git-for-windows/git/wiki/Package-management
- https://github.com/git-for-windows/git/wiki/Building-Git
- https://github.com/git-for-windows/git/wiki/Building-msys2-runtime

**Comment by dscho** [comment]
Oh, and of course https://github.com/git-for-windows/git/wiki for a general overview.

**Comment by noinkling** [comment]
> You will also need to set HOME and TERM, at least. And maybe adjust the PATH.
What about `EXEPATH` and `PLINK_PROTOCOL`, are they important?

Path appeared to be the same between `git-bash.exe` and ConsoleZ with `usr\bin\sh.exe --login -i`, apart from a duplicate `/usr/bin` and `/mingw64/bin` in Mintty.

**Comment by cristovaov** [comment]
..and following up: without any other setting apart from what I mentionned above - `$HOME` is already set to my user profile and `$TERM` returns 'cygwin' instead of 'xterm'. [Some info on the difference.] (http://www.pitt.edu/~naraehan/ling2050/config-cygwin.html)

@dscho I haven't got round yet to it (my win10 box is ready to go though) Thank you for the links!
Looking forward to a reply for @noinkling 's question.

**Comment by dscho** [comment]
Just to steer this discussion back to a healthy path: *not* calling `/bin/bash.exe --login -i` is probably the worst decision here. The best decision would be to fix `/bin/bash.exe --login -i` to do what we want it to do.

I *could* imagine that a command-line option to change the set of flags [passed to `CreateProcessW()`](https://github.com/git-for-windows/git/blob/a6803cd62bba2238d6d42115f6db9d8e22811a54/compat/win32/git-wrapper.c#L549-L590) would fix it. But this needs careful analysis and testing instead of wild west style tampering with environment variables. It would be much better to actually spend time to solve the problem rather than to introduce a new can of problems.

Of course, you are free to continue in the direction of avoiding `/bin/bash.exe --login -i` in favor of the environment variable hackery, but in that case this ticket would no longer be about Git for Windows now, would it.

**Comment by dlong500** [comment]
So is this PR dead now?  Any hope of actually fixing the problem faced by ConsoleZ users?

**Comment by dscho** [comment]
@dlong500 have you read [my comment](https://github.com/git-for-windows/git/pull/205#issuecomment-138485999)?

**Comment by dlong500** [comment]
@dscho Yes, but that seems pretty open-ended with no path to resolution other than suggesting some work that could eventually be done.  Not meaning to be critical; just simply wanting to know if there was going to be any progress made in the immediate future.

**Comment by bviktor** [comment]
@dlong500 I ended up patching each and every Git for Windows release with the patch above. Quite ridiculous indeed.

**Comment by dlong500** [comment]
I understand that this PR may not work if it breaks other things.  What I don't understand is exactly what seems to have changed in the new version of bash that is causing the problem since older versions worked fine with ConsoleZ.  At this point I think I just have to move on and try something else like ConEmu since I don't have the time to try to fix my editor.

**Comment by dscho** [comment]
> What I don't understand is exactly what seems to have changed in the new version of bash that is causing the problem since older versions worked fine with ConsoleZ

We used to have a script that starts Bash after setting up a couple of environment variables. Now we have a proper `.exe`, but that seems to interfere with the way ConsoleZ expects things to be.

**Comment by dscho** [comment]
> @dlong500 I ended up patching and rebuilding each and every Git for Windows release for my company's internal use with the patch above. Quite ridiculous indeed.

@bviktor your patch breaks *my* setup. And I am responsible for many, many more installations in addition. I did propose a solution, did I not? A command-line option that works around the issue until the time when some developer comes along who knows how to fix this issue better than I could.

**Comment by bviktor** [comment]
There have been several pages of discussion, so again, what is your workaround?

**Comment by dscho** [comment]
> There have been several pages of discussion, so again, what is your workaround?

To introduce a command-line option to make your change opt-in rather than force-in. BTW I tried to spare you the time to read the pages, and also the time to ask again, by stating:

> A command-line option that works around the issue until the time when some developer comes along who knows how to fix this issue better than I could.

:wink:

**Comment by bviktor** [comment]
The only reference I could find is this:

> I think that we should introduce a new command-line option to install that ctrl handler and not do that by default

Never was it explained if it was actually implemented, and if so, how do I use it.

**Comment by draenor** [comment]
A command-line option to opt-in sounds great. Is it merely a problem of implementation or is the debate to find a more sustainable approach?

**Comment by dscho** [comment]
> A command-line option to opt-in sounds great. Is it merely a problem of implementation or is the debate to find a more sustainable approach?

@draenor the optimal solution would not require an opt-in, but also would not break my setup ;-)

> > I think that we should introduce a new command-line option to install that ctrl handler and not do that by default
> 
> Never was it explained if it was actually implemented, and if so, how do I use it.

We are currently discussing as comments to a Pull Request. This Pull Request wants to introduce a Ctrl handler but unfortunately that handler breaks my setup, as I [described earlier](https://github.com/git-for-windows/git/pull/205#issuecomment-113185246). The command-line option I was talking about would obviously *not* install that Ctrl handler all the time, but only when asked for.

But really, ideally I would hope for somebody to come up with a proper fix, based on a much deeper understanding of console issues than I possess.

**Comment by bviktor** [comment]
So basically, previously you were asking me why I don't use the workaround that doesn't exist yet?

**Comment by dscho** [comment]
@bviktor I was asking everybody who needs this issue fixed to work on this Pull Request. In particular, as long as it breaks already *my* setup, it should be clear that it cannot be merged. It would be a lousy idea to accept Pull Requests that make the project's maintainer's own installation unusable. That actually should go without saying.

Now, we have a diff in this Pull Request. I made the suggestion to guard the newly-introduced code by an opt-in command-line option. This would serve as a work-around until the problem gets fixed properly, without breaking anybody's setup by way of fixing the described problem.

This is where this Pull Request is at, and where it has been for quite a while now. I do not believe that talking about this more will help anything, so I will stop for now.

**Comment by rimrul** [comment]
From a short websearch[1,2,3,4] the `CtrlHandler` should propably look more like this:

```C
    /*
    Control Handler Function
    Ctrl+C is disabled!
    */
    BOOL WINAPI CtrlHandler( DWORD fdwCtrlType ) 
    {
        return fdwCtrlType == CTRL_C_EVENT;
    }
```

Maybe there is an need to also Handle `CTRL_BREAK_EVENT`.

--

[1][Adam Sawicki - Handling Ctrl+C in Windows Console Application](http://asawicki.info/news_1465_handling_ctrlc_in_windows_console_application.html)
[2][Stackoverflow - Handle CTRL+C on Win32](http://stackoverflow.com/questions/18291284/handle-ctrlc-on-win32)
[3][MSDN - SetConsoleCtrlHandler function](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686016%28v=vs.85%29.aspx)
[4][MSDN - HandlerRoutine callback function](https://msdn.microsoft.com/en-us/library/windows/desktop/ms683242%28v=vs.85%29.aspx)

**Comment by cbucher** [comment]
There are various ways to use it.
You can choose to handle only CTRL-C or all stop events. If you choose to handle all events you are sure the shell will terminate before the "redirector".

I think CTRL-Break should be handled too. cmd or bash don't terminate when you press CTRL-Break.

**Comment by cbucher** [comment]
@dscho >In particular, as long as it breaks already my setup, it should be clear that it cannot be merged. 

What is broken with this patch?

**Comment by rimrul** [comment]
>If you choose to handle all events you are sure the shell will terminate before the "redirector".
I think CTRL-Break should be handled too.

I agree, but I think we shouldn't just mark `CTRL_CLOSE_EVENT` as handled without doing anything. How about this?

```C
return fdwCtrlType != CTRL_CLOSE_EVENT;
```

>>In particular, as long as it breaks already my setup, it should be clear that it cannot be merged.

>What is broken with this patch?

That seems to be an important question here on our way to "Why is that broken?". One answer apppears in the discussion above:

>if I run the patches git-cmd.exe and try to interrupt a long-running git rev-list HEAD, it simply freezes. With no way to recover other than calling the Task Manager, guessing the correct git.exe and terminating the process (with two frightening warnings).

This is definitly an issue, but maybe, just maybe this code isn't the cause, but something that triggers the symptoms?

**Comment by dscho** [comment]
> What is broken with this patch?

https://github.com/git-for-windows/git/pull/205#issuecomment-113185246

**Comment by dscho** [comment]
(I'll stop doing my famous imitation of a broken record and bow out of the discussion for a while.)

**Comment by bviktor** [comment]
So basically now you expect someone else to fix a problem for you that only you reproduced so far? Mmkay.

Maybe if you provided an _exact_ test case (something that you always expect from others) instead of vague complaints, that would actually speed thing up a little.

**Comment by dscho** [comment]
@bviktor nice try. Fortunately, I am way too busy with preparing the 2.6.0 release, so I cannot take your bait.

**Comment by cbucher** [comment]
>Unfortunately, my concerns were legitimate: if I run the patches git-cmd.exe and try to interrupt a long-running git rev-list HEAD, it simply freezes. With no way to recover other than calling the Task Manager, guessing the correct git.exe and terminating the process (with two frightening warnings).

- the *redirector* has **no** interaction with bash (*redirector* is an external process which is just waiting signal of bash termination)
- the *redirector* receives the CTRL-C event **after** bash processed this event (you can add some *printf* too see how the chain is working)
- and bash receives CTRL-C event **after** git  processed this event.

In your problem description, git receives and processes the CTRL-C event in *first*!
So if *git* don't respond to CTRL-C, the problem should be in git process, and not in external processes! (I'm not saying that the git code is involved)

I tested CPU load programs and I have no problem with CTRL-C (with or without the patch).

**Comment by dscho** [comment]
Sorry, there was a typo. I meant to say the patche*d* `git-cmd.exe`. The Git wrapper serves also as `git-cmd.exe`, `git-bash.exe`, `git-gui.exe`, `gitk.exe` and even as the redirectory for the legacy `/bin/bash.exe`.

**Comment by dscho** [comment]
To see the problem, just rebuild the `mingw-w64-x86_64-git` package as described [here](https://github.com/git-for-windows/git/wiki/Package-management#rebuild-packages) and with the applied patch. Then start `/cmd/git-cmd.exe`.

**Comment by cbucher** [comment]
I cannot reproduce your problem with current binaries and patched git-cmd.exe ... 
Why should I rebuild package to see the problem?

**Comment by dscho** [comment]
> Why should I rebuild package to see the problem?

Because the code your patch modifies *affects* `git-cmd.exe`, as built in that package.

**Comment by theyankee76** [comment]
The issue is still present in git for windows 2.6.2. Any plans to include the bugfix in a new release ?

**Comment by dscho** [comment]
> Any plans to include the bugfix in a new release ?

@theyankee76 please read through the comments above. It is a bad idea to include bug fix that introduces a new bug.

**Comment by cbucher** [comment]
@dscho > It is a bad idea to include bug fix that introduces a new bug.

waiting for a proof ...
Is someone else can reproduce your issue?

I explain why this assertion can only be wrong.

**Comment by theyankee76** [comment]
I don't know if it can help you but I made several tries, and I have noticed with ConsoleZ:
- with shell set to: "C:\devel\git2\git-cmd.exe --command=usr\bin\bash.exe --login -i", when pressing CTRL-C or CTRL-D, the current process is killed, but the tab also.
- with shell set to: "C:\devel\git2\usr\bin\bash.exe --login -i", when calling a long process (a maven build for example), my computer hangs up, and I got too many instances of "bash.exe" in a tree.

image

**Comment by dscho** [comment]
> waiting for a proof

Look, I reported a problem with this Pull Request because I already have proof that it breaks things here. If you want this Pull Request to go anywhere, you should see for yourself that [the bug I reported](https://github.com/git-for-windows/git/pull/205#issuecomment-113185246) is still there. You never updated anything in this Pull Request in response to my concerns. That is why, after months of waiting, I closed the Pull Request without merging it.

**Comment by cbucher** [comment]
You always respond by a link to that:

>Unfortunately, my concerns were legitimate: if I run the patches git-cmd.exe and try to interrupt a long-running git rev-list HEAD, it simply freezes. With no way to recover other than calling the Task Manager, guessing the correct git.exe and terminating the process (with two frightening warnings).

This is not a proof. This is just a vague description of two events, you must prove the causality.
Where are the dumps, the processes chain, etc...

I explain why the causality was questionable:

>the redirector has no interaction with bash (redirector is an external process which is just waiting signal of bash termination)
>the redirector receives the CTRL-C event after bash processed this event (you can add some printf too see how the chain is working)
>and bash receives CTRL-C event after git processed this event.
>In your problem description, git receives and processes the CTRL-C event in first!

I can say:
- "I have a headache"
- "II slept with my shoes"
But I cannot say: "Sleeping with shoes causes a headache"

**Comment by cbucher** [comment]
/bin/git.exe is a redirector too... and I assume that's use the same source file `git-wrapper.cpp`.
In this case this is logical that /bin/git.exe is not killed by CTRL-C with this patch.

In this case, I let you fix the problem... (it's your project !)

How to handle CTRL-C:
- shells should terminate
- applications should terminate

**Comment by theyankee76** [comment]
Don't hesitate if I can help you and make any tests to solve this issue.

**Comment by kjeremy** [comment]
:+1:

**Comment by yoshiwaan** [comment]
Is there information available about what the uses of those environment variables set using the redirector libary are (other than adding git to PATH)?

**Comment by dscho** [comment]
Yes, see the source code of the wrapper: https://github.com/git-for-windows/git/blob/master/compat/win32/git-wrapper.c

**Comment by dscho** [comment]
But note that the preferred way to go forward would be to *solve* this problem, [as I indicated earlier](https://github.com/git-for-windows/git/pull/205#issuecomment-143801072).

**Comment by kjeremy** [comment]
@dscho can you summarize what your problem is with this patch?  It's a little unclear after reading through the traffic (there are a lot of comments).  I understand you want a command line switch but I do not understand the specifics of why.

**Comment by dscho** [comment]
@kjeremy yes, there as been a lot of chatter here.

The patch disables a workaround I introduced to fix *another* bug. Applying the patch straight, as-is, would re-introduce the bug. This is doubly bad in this instance because this other bug is a bug *I* experience, while I do not use ConsoleZ and therefore would not even benefit from this patch, were I to merge this PR as-is.

**Comment by Maximus5** [comment]
@dscho Can you post a link to this *another* re-introduced bug?

@cbucher Actually, ConsoleZ has inherited bad design from console. It's wrong to rely on termination of the "root" process. Example? Console application is started, downloads an update from internet, starts new version **in the current console** and terminates. What would happen with ConsoleZ? It would be terminated too most probably, regardless the fact there are running console processes.

**Comment by dscho** [comment]
> @dscho Can you post a link to this another re-introduced bug?

Sorry, it's been too long, I would have to search myself where I introduced that line that this patch wants to disable (IIRC).

**Comment by bviktor** [comment]
I keep patching `git-cmd.exe` for our internal use. We have about 35 developers and have been using the patched builds happily ever since. I've created a page with the builds so that others don't have to waste time with patching and building:

### [Git CMD patched for ConsoleZ](http://vault-tec.info/post/139474607261/git-cmd-patched-for-consolez)

Cheers.

**Comment by dscho** [comment]
@bviktor that might seem all nice and dandy for you, but it is *still not a proper bug fix*.

**Comment by bviktor** [comment]
I hope you realize that what you've been saying for the past 8 months is basically:

> This patch breaks something, somewhere, but I don't know and/or I won't tell.

Of course, it's up to you, do as you like, I'm just stating the facts.

In the meanwhile we want to use Git and ConsoleZ without accidentally closing the terminal all the time. As long as it works for us (and it does, we transfer several hundred megabytes and several hundred changesets a day over Git), I couldn't care less about your phantom regressions.

**Comment by dscho** [comment]
And yes, I realize that I have a typo in my observation. It should have talked about a "patched `git-cmd.exe`", not a "patches `git-cmd.exe`". Maybe it was that typo that let everybody simply ignore the pretty valid concerns.

So here goes my challenge: *do* something about it. Talking about it will get us exactly nowhere, as it has been proven beyond doubt by now.

Thanks.

**Comment by dscho** [comment]
> I couldn't care less about your phantom regressions.

Let us drop such despicably patronizing comments.

Back to *work*, yes?

**Comment by cbucher** [comment]
@Maximus5 
> Actually, ConsoleZ has inherited bad design from Console2. It's wrong to rely on termination of the "root" process. Example? Console application is started, downloads an update from internet, starts new version in the current console and terminates. What would happen with ConsoleZ? It would be terminated too most probably, regardless the fact there are running console processes.

I don't agree!
Each program that launches an executable and monitors it has the same issue.
There is no reason to hook every process!
The redirector can be very easily fixed: I provided the correction a long time ago...
The redirector is not backward compatible with previous batch files.

**Comment by Maximus5** [comment]
@cbucher I don't refute that git-cmd must survive on Ctrl-C. On the other hand, ConsoleZ definitely has a problem. Example: Far Manager update procedure works in standard Windows terminal, works in ConEmu, but ConsoleZ just closes. My own opinion - it's better to patch both git-cmd and ConsoleZ. But I don't care a lot about it.

**Comment by cbucher** [comment]
@Maximus5 The workaround is easy: start the application from a shell that will not terminate.
For example: `cmd.exe /K "..."`

This workaround works with git-cmd too.

But git-cmd monitors and waits the termination of the real bash. In this case, the termination with CTRL-C is abnormal and should be fixed.

**Comment by Maximus5** [comment]
@cbucher Obviously that is not a workaround at all. After that **both** applications would read console input buffer, because "cmd.exe /k" returns to prompt and new Far.exe starts in "background". Anyway, this is offtopic here.

**Comment by yoshiwaan** [comment]
Please, let me at least try to steer this back into the direction of a proper fix. There's been some flat out rudeness in this thread which is not helping things, so please keep it civil.

People need to remember that Git for Windows is a packaged product with many users and the ones that want to use ConsoleZ are probably a minority and it's an optional extra at that. The onus is on the ConsoleZ developers and users to fix this issue, not @dscho. 

The idea that this fix should be merged despite re-introducing a regression is ludicrous and should never happen. The onus isn't on @dscho to prove this code works, it's on @cbucher as the owner of the PR. If @dscho is saying it doesn't work and it affects the core product then it's never going to get merged. 

That being said without the information from @dscho to reproduce the regression from this PR then there's nothing further that can be done by anyone to help this process. So @dscho please provide as detailed steps on how to reproduce the regression you encountered (including how you built the program and what platform it was on) so that we can try to reproduce it. Better yet write an automated test for it that breaks on your machine with the patch but works without it.
It would also be EXTREMELY valuable if you can provide the bug that this regression re-introduces. I know it's work for you but you are one of the stewards of this project and thus you have the knowledge to find it as you worked on it.

Once we have that information then the goal of this PR should be changed to have a fix that doesn't break mainline and we, the users of ConsoleZ, should work to do that.

**Comment by cbucher** [comment]
@Maximus5 
>After that both applications would read console input buffer, because "cmd.exe /k" returns to prompt and new Far.exe starts in "background". Anyway, this is offtopic here.

If after the update, you get two separate consoles, then there is no reason to choose one over the other.
In short you write specific code to support Far Manager.
Personally it's not my philosophy

**Comment by dscho** [comment]
@yoshiwaan thanks for your thoughtful comment.

While I was unable to reproduce [my original issue (`rev-list HEAD` would not terminate with Ctrl+C)](https://github.com/git-for-windows/git/pull/205#issuecomment-113185246), I found the culprit of all those woes: the Console Ctrl Handler installed by this Pull Request also prevented "close window" events from being handled, among other things.

I fixed these issues, patched up the commit message, tested *the heck* out of it, and pushed the fix as b85db5b1b931ccbbfa3468b2ea02af10cd34b1a9. Even if almost nothing of the original patch remained in place, I kept @cbucher as the author, to give him credit.

**Comment by yoshiwaan** [comment]
@dscho Awesome, thanks. So this will go out with the next release?

**Comment by dscho** [comment]
> this will go out with the next release?

Yes.

**Comment by bviktor** [comment]
Confirmed, thanks.

**Comment by redemption** [comment]
Git for Windows 2.7.2 was just released a few hours ago, I downloaded it and can confirm the CTRL-C issue is no longer a problem with Console 2 (the precursor to ConsoleZ).  Thanks very much for the fix, was driving me and our team mad!

**Comment by noinkling** [comment]
Thanks for fixing this, and thanks for all your work on git-for-windows in general @dscho. I know you received some pretty rude and self-entitled comments over this issue (in other threads as well as this one).

**Comment by dscho** [comment]
@noinkling thank you so much for your kind words. You have no idea what this means to me.

**Comment by avindra** [comment]
You all are saints for those of us stuck on Windows at work. Thank you! :+1:

**Comment by FredoVelcro** [comment]
Thanks a lot for the fix ! GREAT !

**Comment by hughwphamill** [comment]
Came across this PR while researching this issue, looks like it's already fixed, just want to echo the 'thanks for the hard work' of the commenters above.. off to download latest & greatest now...

:smile:

**Comment by msalahz** [comment]
:+1:  Thanks Everyone.
